http-client-jdk/src/main/java/io/micronaut/http/client/jdk/DefaultJdkHttpClientRegistry.java

```diff
@@ -0,0 +1,328 @@
+        return getJsonCodec(beanContext).cloneWithFeatures(jsonFeatures);
+    }
+
+    private static MapperMediaTypeCodec getJsonCodec(BeanContext beanContext) {
+        return beanContext.getBean(MapperMediaTypeCodec.class, Qualifiers.byName(MapperMediaTypeCodec.REGULAR_JSON_MEDIA_TYPE_CODEC_NAME));
+    }
+
+    /**
+     * Creates a {@literal java.net.http.*} HTTP Client.
+     *
+     * @param injectionPoint
+     * @param loadBalancer
+     * @param configuration
+     * @param beanContext
+     * @return A {@literal java.net.http.*} HTTP Client
+     */
+    @Bean
+    @BootstrapContextCompatible
+    @Primary
+    protected HttpClient httpClient(
+        @Nullable InjectionPoint<?> injectionPoint,
+        @Parameter @Nullable LoadBalancer loadBalancer,
+        @Parameter @Nullable HttpClientConfiguration configuration,
+        BeanContext beanContext
+    ) {
+        return resolveDefaultHttpClient(injectionPoint, loadBalancer, configuration, beanContext);
+    }
+
+    private HttpClient resolveDefaultHttpClient(
+        @Nullable InjectionPoint<?> injectionPoint,
+        @Nullable LoadBalancer loadBalancer,
+        @Nullable HttpClientConfiguration configuration,
+        BeanContext beanContext
+    ) {
+        if (loadBalancer != null) {
+            if (configuration == null) {
+                configuration = defaultHttpClientConfiguration;
+            }
+            return buildClient(
+                loadBalancer,
+                null,
+                configuration,
+                null,
+                loadBalancer.getContextPath().orElse(null),
+                beanContext
+            );
+        } else {
+            return getClient(injectionPoint != null ? injectionPoint.getAnnotationMetadata() : AnnotationMetadata.EMPTY_METADATA);
+        }
+    }
+
+    @Override
+    public HttpClient getClient(AnnotationMetadata annotationMetadata) {
+        final ClientKey key = getClientKey(annotationMetadata);
+        return getClient(key);
+    }
+
+    private ClientKey getClientKey(AnnotationMetadata metadata) {
+        HttpVersionSelection httpVersionSelection = HttpVersionSelection.forClientAnnotation(metadata);
+        String clientId = metadata.stringValue(Client.class).orElse(null);
+        String path = metadata.stringValue(Client.class, "path").orElse(null);
+        List<String> filterAnnotation = metadata
+            .getAnnotationNamesByStereotype(FilterMatcher.class);
+        final Class<?> configurationClass =
+            metadata.classValue(Client.class, "configuration").orElse(null);
+        JsonFeatures jsonFeatures = jsonMapper.detectFeatures(metadata).orElse(null);
+
+        return new ClientKey(httpVersionSelection, clientId, filterAnnotation, path, configurationClass, jsonFeatures);
+    }
+
+    private HttpClient getClient(ClientKey key) {
+        return clients.computeIfAbsent(key, clientKey -> {
```

Comment:
do we do this similarly in the netty client?

computeIfAbsent can lead to deadlocks if it is called recursively, so it seems a bit dangerous to have such a big lambda. but maybe it's fine.

Reply:
Yes

https://github.com/micronaut-projects/micronaut-core/blob/4ec18d886a0c705b5f2ea91516dfaee95644cbcc/http-client/src/main/java/io/micronaut/http/client/netty/DefaultNettyHttpClientRegistry.java#L310

It's 9 lines longer for Netty

Reply:
ok, then just ignore it :D